synapseclient/table.py

```diff
@@ -829,75 +857,155 @@ class Dataset(SchemaBase):
+            if not self.has_item(dataset_item.get('entityId')):
+                self.properties.datasetItems.append(dataset_item)
+            else:
+                if force:
+                    self.remove_item(dataset_item.get('entityId'))
+                    self.properties.datasetItems.append(dataset_item)
+                else:
+                    raise ValueError(
+                        f"Duplicate item found: {dataset_item.get('entityId')}. "
+                        "Set force=True to overwrite the existing item.")
         else:
             raise ValueError("Not a DatasetItem? %s" % str(dataset_item))
 
-    def add_items(self, dataset_items: List[Dict[str, str]]):
+    def add_items(self, dataset_items: List[Dict[str, str]], force: bool = False):
         """
-        :param dataset_items: a list of dataset items
+        :param dataset_items:   a list of dataset items
+        :param force:           force add items
         """
         for dataset_item in dataset_items:
-            self.add_item(dataset_item)
+            self.add_item(dataset_item, force)
 
     def remove_item(self, item_id: str):
         """
         :param item_id: a single dataset item Synapse ID
         """
         item_id = id_of(item_id)
         if item_id.startswith("syn"):
-            for i, curr_item in enumerate(self.properties.dataset_items):
+            for i, curr_item in enumerate(self.properties.datasetItems):
                 if curr_item.get('entityId') == item_id:
-                    del self.properties.dataset_items[i]
+                    del self.properties.datasetItems[i]
                     break
         else:
             raise ValueError("Not a Synapse ID: %s" % str(item_id))
 
+    def empty(self):
+        self.properties.datasetItems = []
+
     def has_item(self, item_id):
         """
         :param item_id: a single dataset item Synapse ID
         """
-        return any(item['entityId'] == item_id for item in self.properties.dataset_items)
+        return any(item['entityId'] == item_id for item in self.properties.datasetItems)
+
+    def add_folder(self, folder_id: str, force: bool = False):
+        """
+        :param folder_id:   a single Synapse Folder ID
+        :param force:       force add items from folder
+        """
+        if not self.__dict__.get('folders_to_add', None):
+            self.__dict__['folders_to_add'] = set()
+        self.__dict__['folders_to_add'].add(folder_id)
+        if self.force != force:
+            self.force = force
+
+    def add_folders(self, folder_ids: List[str], force: bool = False):
+        """
+        :param folder_ids:  a list of Synapse Folder IDs
+        :param force:       force add items from folders
+        """
+        if isinstance(folder_ids, list) or isinstance(folder_ids, set) or \
+                isinstance(folder_ids, tuple):
+            self.force = force
```

Comment:
Why add `force` as a class attribute?

Reply:
add_folder() and add_folders() do not actually add files into the dataset; it simply just tells the object what folders will need to be added.

The method that actually adds the file is the hidden helper function, _add_folder_files, which will need some way of knowing whether files should be overwritten.  Let me know if there's a better way to go around it.

Reply:
I see. Currently in the web client, the default is that dataset items are replaced so `force=True` by default.

Reply:
Thinking it over, perhaps we should continue to offer the option to turn off `force` in the methods? I know we should be emulating the web UI's behavior, but if it were me and I was programmatically working with datasets, I think it'll be a convenient option to have.  Otherwise, the alternative would be to write a method that would turn on/off force-add, and requiring users to call that before adding any items.  If it were me, I would find that super inconvenient, as opposed to just doing `add_item(..., force=False)`.

Of course, another option is to not allow changing this behavior at all. That is, all duplicate items will overwrite the existing ones.  And if a user doesn't want that to happen, then they should use `has_item()` beforehand.  My only issue with this is that since the `add_item()` method already calls on `has_item()`, it'll be a little redundant (of course, the user wouldn't know that..... but we would :sweat_smile:)